src/main/scala/org/apache/sysml/compiler/lang/source/DML.scala

```diff
@@ -0,0 +1,473 @@
+
+        val alg = new Source.Algebra[D] {
+
+          def empty: D = env => ""
+
+          // Atomics
+          def lit(value: Any): D = env => value match {
+            //@formatter:off
+            case value: Boolean => if (value) "TRUE" else "FALSE"
+            case value: Char => s""""${value}""""
+            case value: String => s""""${escape(value)}""""
+            case null => "null"
+            case value: Unit => ""
+            case _ => value.toString
+            //@formatter:on
+          }
+
+          def this_(sym: u.Symbol): D = env => {
+            s"_this: ${sym.name.decodedName.toString}"
+          }
+
+          def bindingDef(lhs: u.TermSymbol, rhs: D): D = valDef(lhs, rhs)
+
+          def ref(target: u.TermSymbol): D = ???
+
+          def loop(cond: D, body: D): D = whileLoop(cond, body)
+
+          def moduleAcc(target: D, member: u.ModuleSymbol): D = ???
+
+          override def bindingRef(sym: u.TermSymbol): D = env => {
+            printSym(sym)
+          }
+
+          override def moduleRef(target: u.ModuleSymbol): D = env =>
+            printSym(target)
+
+          // Definitions
+          override def valDef(lhs: u.TermSymbol, rhs: D): D = env =>
+            s"${printSym(lhs)} = ${rhs(env)}"
+
+          override def parDef(lhs: u.TermSymbol, rhs: D): D = env => {
+            val l = lhs.name.decodedName
+            val tpe = lhs.typeSignature
+            val r = rhs(env)
+
+            if (r.isEmpty) {
+              s"$l"
+            } else {
+              s"$l = $r"
+            }
+          }
+
+          // Other
+
+          /** type ascriptions such as `var x: Int` */
+          def typeAscr(target: D, tpe: u.Type): D = env => "" // TODO check for types allowed in SystemML
+
+          def defCall(target: Option[D], method: u.MethodSymbol, targs: Seq[u.Type], argss: Seq[Seq[D]]): D = env => {
+            val s = target
+            val args = argss flatMap (args => args map (arg => arg(env)))
+
+            (target, argss) match {
+
+              /* matches tuples */
+              case (Some(tgt), _) if isApply(method) && api.Sym.tuples.contains(method.owner.companion) =>
+                ""
+
+              case (Some(tgt), _) if method == api.Sym.foreach || method.overrides.contains(api.Sym.foreach) =>
+                forLoop(tgt, targs, argss.flatten, env)
+
+              /* matches unary methods without arguments, e.g. A.t */
+              case (Some(tgt), Nil) if isUnary(method) =>
+                s"${printSym(method)}${tgt(env)}"
+
+              /* matches apply methods with one argument */
+              case (Some(tgt), ((arg :: Nil) :: Nil)) if isApply(method) => {
+                val module = tgt(env)
+
+                if (module == "Vector") { // Vector.apply(Array(...))
+                  printConstructor(method, argss, env, true)
+                }
+
+                else {
+                  s"${tgt(env)}${printMethod(" ", method, " ")}${arg(env)}"
```

Comment:
Is this logic correct for every object the user might call `apply` on, or should there be some additional error checking code here?